core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+		let import_block = BlockImportParams {
+			origin,
+			header: checked_header,
+			post_digests: vec![seal],
+			body,
+			finalized: false,
+			justification,
+			auxiliary: vec![(key, Some(aux.encode()))],
+			fork_choice: ForkChoiceStrategy::Custom(aux.total_difficulty > best_aux.total_difficulty),
+		};
+
+		Ok((import_block, None))
+	}
+}
+
+/// Register the PoW inherent data provider, if not registered already.
+fn register_pow_inherent_data_provider(
+	inherent_data_providers: &InherentDataProviders,
+) -> Result<(), consensus_common::Error> {
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		inherent_data_providers
+			.register_provider(srml_timestamp::InherentDataProvider)
+			.map_err(Into::into)
+			.map_err(consensus_common::Error::InherentData)
+	} else {
+		Ok(())
+	}
+}
+
+/// The PoW import queue type.
+pub type PowImportQueue<B> = BasicQueue<B>;
+
+/// Import queue for PoW engine.
+pub fn import_queue<B, C, Algorithm>(
+	block_import: BoxBlockImport<B>,
+	client: Arc<C>,
+	algorithm: Algorithm,
+	inherent_data_providers: InherentDataProviders,
+) -> Result<PowImportQueue<B>, consensus_common::Error> where
+	B: BlockT<Hash=H256>,
+	C: ProvideRuntimeApi + HeaderBackend<B> + BlockOf + ProvideCache<B> + AuxStore,
+	C: Send + Sync + AuxStore + 'static,
+	C::Api: BlockBuilderApi<B>,
+	Algorithm: PowAlgorithm<B> + Send + Sync + 'static,
+{
+	register_pow_inherent_data_provider(&inherent_data_providers)?;
+
+	let verifier = PowVerifier {
+		client: client.clone(),
+		algorithm,
+		inherent_data_providers,
+	};
+
+	Ok(BasicQueue::new(
+		verifier,
+		block_import,
+		None,
+		None
+	))
+}
+
+/// Start the background mining thread for PoW. Note that because PoW mining
+/// is CPU-intensive, it is not possible to use an async future to define this.
+/// However, it's not recommended to use background threads in the rest of the
+/// codebase.
+pub fn start_mine<B: BlockT<Hash=H256>, C, Algorithm, E>(
+	mut block_import: BoxBlockImport<B>,
+	client: Arc<C>,
+	algorithm: Algorithm,
+	mut env: E,
+	preruntime: Vec<u8>,
+	round: u32,
+	inherent_data_providers: inherents::InherentDataProviders,
+) where
+	C: HeaderBackend<B> + AuxStore + 'static,
+	Algorithm: PowAlgorithm<B> + Send + Sync + 'static,
+	E: Environment<B> + Send + Sync + 'static,
+	E::Error: std::fmt::Debug,
+{
+	if let Err(_) = register_pow_inherent_data_provider(&inherent_data_providers) {
+		warn!("Registering inherent data provider for timestamp failed");
+	}
+
+	thread::spawn(move || {
+		loop {
+			match mine_loop(
+				&mut block_import,
+				client.as_ref(),
+				&algorithm,
+				&mut env,
+				&preruntime,
+				round,
+				&inherent_data_providers
+			) {
+				Ok(()) => (),
+				Err(e) => error!(
+					"Mining block failed with {:?}. Sleep for 1 second before restarting...",
+					e
+				),
+			}
+
+			std::thread::sleep(std::time::Duration::new(1, 0));
+		}
+	});
+}
+
+fn mine_loop<B: BlockT<Hash=H256>, C, Algorithm, E>(
```

Comment:
I do not see how cancellation is handled in this design. We have several cases where we want to cancel mining early:
- A new best block had been received and we should restart mining on top of it
- The system is shutting down

Reply:
Cancellation is handled like this: for `PowAlgorithm::mine`, it takes a parameter called `round`. The CPU miner is then expected to only mine that particular "rounds" as defined by the algorithm. When calling `start_mine`, it is expected that the caller passes an initial `round` that results in subsecond execution of `PowAlgorithm::mine`. If seal is found after rounds, then we import the block. If no seal is found, then `mine` returns `None`. The mining function then loops, checks whether best block has changed. If not, directly call `PowAlgorithm::mine` again. Otherwise, build a new block and mining on that instead.

TBH I'm hoping to keep `Algorithm` as the generic type name. `B`, `C`, `E` all has standard meanings in other consensus engines, but `Algorithm` is one specific to PoW engine. Keeping it like this just makes things more descriptive, and it doesn't actually result in long code anyway.